### ISSUE TITLE
Make ImageViewTarget an open class

### ISSUE THREAD

**Issue by chrisbanes** [coil-kt/coil#98]
Currently ImageViewTarget is final, meaning it can't be extended. It would be useful to make it `open`.

*Why?*
I want to add a custom animation/transition when the image is loaded. I could add a transformation and wrap the drawable in a custom `Animatable` (although no idea if that would work), but really I just want extend `ImageViewTarget` and add something similar to its crossfade functionality.

**Comment by colinrtwhite** [comment]
Sorry for the late reply on this. `ImageViewTarget` is currently closed since I didn't want consumers relying on the internal behaviour of `onSuccessCrossfade`.

`PoolableViewTarget` requires implementations to stop using the previous `Drawable` as soon as the next lifecycle method is called. This works great when using `setImageDrawable`, but it doesn't account for transitions that need to use both Drawables for the duration of a transition. Currently, `ImageViewTarget` works around this with some special internal behaviour in `onSuccessCrossfade`. This is definitely a shortcut and I want to make this behaviour public, but haven't settled on a design yet.

I want to promote [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) so my goal is to expose something like a `TransitionTarget` where implementations that want to handle transitions have access to a transition-friendly `onSuccess`. I don't think it makes sense to add DataSource to the base Target interface since most targets don't need to suspend the `onSuccess` method while animating.

Here's what I'm thinking for `TransitionTarget` (let me know what you think):

```kotlin
interface TransitionTarget : Target {
    suspend fun onSuccess(result: Drawable, source: DataSource)
}
```

Basically, Coil needs a good way to support custom transitions, which it doesn't have at the moment.

**Comment by chrisbanes** [comment]
Sounds good to me. Making the onSuccess() suspend while the transition/animation takes place is neat API.
